### PR TITLE
expression: set json to right type

### PR DESCRIPTION
**Pull request by xiongjiwei**
Signed-off-by: xiongjiwei <xiongjiwei1996@outlook.com>

### What problem does this PR solve?

Issue Number: close https://github.com/pingcap/tidb/issues/37339

Problem Summary:

### Check List

Tests 

- [x] Unit test
- [ ] Integration test
- [ ] Manual test (add detailed scripts or steps below)
- [ ] No code

### Release note

Please refer to [Release Notes Language Style Guide](https://pingcap.github.io/tidb-dev-guide/contribute-to-tidb/release-notes-style-guide.html) to write a quality release note.

```release-note
None
```

### CODE REVIEW

**Review by tangenta — APPROVED** [review]

File: expression/typeinfer_test.go [hunk 1]
```diff
@@ -2976,6 +2975,14 @@ func (s *InferTypeSuite) createTestCase4JSONFuncs() []typeInferTestCase {
 		{"json_merge(c_json, c_json)", mysql.TypeJSON, charset.CharsetUTF8MB4, mysql.BinaryFlag, mysql.MaxBlobWidth, 0},
 		{"json_object('k', 'v')", mysql.TypeJSON, charset.CharsetUTF8MB4, mysql.BinaryFlag, mysql.MaxBlobWidth, 0},
 		{"json_array('k', 'v')", mysql.TypeJSON, charset.CharsetUTF8MB4, mysql.BinaryFlag, mysql.MaxBlobWidth, 0},
+		{"json_pretty(c_json)", mysql.TypeLongBlob, charset.CharsetUTF8MB4, mysql.BinaryFlag, mysql.MaxLongBlobWidth, types.UnspecifiedLength},
```
Comment (xhebox) [inline expression/typeinfer_test.go:2978]: It seems that it only has half of the max length on mysql8:

```
Type:       LONG_BLOB
Collation:  utf8mb4_general_ci (45)
Length:     268435456
```

File: expression/typeinfer_test.go [hunk 2]
```diff
@@ -2965,9 +2965,8 @@ func (s *InferTypeSuite) createTestCase4Literals() []typeInferTestCase {
 
 func (s *InferTypeSuite) createTestCase4JSONFuncs() []typeInferTestCase {
 	return []typeInferTestCase{
-		{"json_type(c_json)", mysql.TypeVarString, charset.CharsetUTF8MB4, 0, 51, types.UnspecifiedLength},
-		// TODO: flen of json_unquote doesn't follow MySQL now.
-		{"json_unquote(c_json)", mysql.TypeVarString, charset.CharsetUTF8MB4, 0, mysql.MaxFieldVarCharLength, types.UnspecifiedLength},
+		{"json_type(c_json)", mysql.TypeVarString, charset.CharsetUTF8MB4, mysql.BinaryFlag, 51, types.UnspecifiedLength},
+		{"json_unquote(c_json)", mysql.TypeLongBlob, charset.CharsetUTF8MB4, mysql.BinaryFlag, mysql.MaxLongBlobWidth, types.UnspecifiedLength},
```
Comment (xhebox) [inline expression/typeinfer_test.go:2969]: Quote and unquote seems to have a decimals of 31 on mysql8.

```
Type:       LONG_BLOB
Collation:  utf8mb4_general_ci (45)
Length:     4294967295
Max_length: 0
Decimals:   31
Flags:      BINARY
```

**Review by xhebox — COMMENTED** [review]
Can not test with mysql5 by ` --column-type-info` since I am on M1. There is no mysql5 arm docker image.

I also noticed `MaxInt` is 21 in mysql8. Maybe it is 20 in mysql5.

File: expression/typeinfer_test.go (hunk 1 above)
Comment (xiongjiwei) [reply expression/typeinfer_test.go]: it is actually `16777216 * 4 * 4`, `16777216` is the max blob length, `*4` for `utf8mb4` max length, and another `*4` for `utf8mb4` to `binary`. It is hard to compatible with MySQL now, I think it is ok to set it to `16777216 * 4`

File: expression/typeinfer_test.go (hunk 2 above)
Comment (xiongjiwei) [reply expression/typeinfer_test.go:2969]: it MySQL, 31 is `DECIMAL_NOT_SPECIFIED` see https://github.com/mysql/mysql-server/blob/fbdaa4def30d269bc4de5b85de61de34b11c0afc/include/decimal.h#L152-L153

**Review by xhebox — APPROVED** [review]

**Pull request merged**